pybricks/util_pb/pb_flash.c

```diff
@@ -0,0 +1,450 @@
+
+// True sets the notCS pin, False resets it
+static void flash_enable(bool enable) {
+    if (enable) {
+        HAL_GPIO_WritePin(GPIOB, GPIO_PIN_12, GPIO_PIN_RESET);
+    } else {
+        HAL_GPIO_WritePin(GPIOB, GPIO_PIN_12, GPIO_PIN_SET);
+    }
+}
+
+// Configure SPI
+static pbio_error_t pb_flash_spi_init(void) {
+
+    // SPI2 parameter configuration
+    hspi2.Instance = SPI2;
+    hspi2.Init.Mode = SPI_MODE_MASTER;
+    hspi2.Init.Direction = SPI_DIRECTION_2LINES;
+    hspi2.Init.DataSize = SPI_DATASIZE_8BIT;
+    hspi2.Init.CLKPolarity = SPI_POLARITY_LOW;
+    hspi2.Init.CLKPhase = SPI_PHASE_1EDGE;
+    hspi2.Init.NSS = SPI_NSS_SOFT;
+    hspi2.Init.BaudRatePrescaler = SPI_BAUDRATEPRESCALER_2;
+    hspi2.Init.FirstBit = SPI_FIRSTBIT_MSB;
+
+    // Save settings
+    if (HAL_SPI_Init(&hspi2) != HAL_OK) {
+        return PBIO_ERROR_IO;
+    }
+
+    // Enable flash
+    flash_enable(true);
+
+    // Command parameters
+    uint32_t timeout = 100;
+    uint8_t CMD_GET_ID = 0x9F;
+    uint8_t id_data[3];
+    HAL_StatusTypeDef err;
+
+    // Send ID command
+    err = HAL_SPI_Transmit(&hspi2, &CMD_GET_ID, sizeof(CMD_GET_ID), timeout);
+    if (err != 0) {
+        return PBIO_ERROR_IO;
+    }
+
+    // Get ID command reply
+    err = HAL_SPI_Receive(&hspi2, id_data, sizeof(id_data), timeout);
+    if (err != 0) {
+        return PBIO_ERROR_IO;
+    }
+    flash_enable(false);
+
+    // Verify flash device ID
+    if (id_data[0] != 239 || id_data[1] != 64 || id_data[2] != 25) {
+        return PBIO_ERROR_NO_DEV;
+    }
+
+    return PBIO_SUCCESS;
+}
+
+static HAL_StatusTypeDef flash_status_read(uint8_t *status) {
+
+    uint32_t timeout = 100;
+
+    // Read status command
+    uint8_t command = 0x05;
+
+    // Write the read status command
+    flash_enable(true);
+    HAL_StatusTypeDef err = HAL_SPI_Transmit(&hspi2, &command, sizeof(command), 100);
+    if (err != HAL_OK) {
+        return err;
+    }
+
+    // Receive data
+    err = HAL_SPI_Receive(&hspi2, status, 1, timeout);
+    flash_enable(false);
+    return err;
+}
+
+static HAL_StatusTypeDef flash_wait_ready() {
+    uint8_t status = 0x03;
+    HAL_StatusTypeDef err;
+    uint32_t start_time = mp_hal_ticks_ms();
+
+    // While write enabled and / or busy, wait
+    while (status & 0x03) {
+        err = flash_status_read(&status);
+        if (err != HAL_OK) {
+            return err;
+        }
+        if (mp_hal_ticks_ms() - start_time > 100) {
+            return HAL_TIMEOUT;
+        }
+        MICROPY_EVENT_POLL_HOOK;
+    }
+    return HAL_OK;
+}
+
+static HAL_StatusTypeDef flash_read(uint32_t address, uint8_t *buffer, uint32_t size) {
+
+    uint32_t timeout = 100;
+
+    // First 1MiB is reserved for firmware updates with official firmware.
+    address += 1024 * 1024;
+
+    uint8_t address_bytes[5];
+
+    // Read command
+    address_bytes[0] = 0x13;
+
+    // Address as big endian
+    address_bytes[1] = (address & 0xff000000) >> 24;
```

Comment:
Applying bit masks is not necessary here (increases firmware size). Assigning to a byte takes care if it already.

Reply:
could be nice to make a `pbio_set_uint16_be()` function too